File: Projects/Offensive/Rootkit practice/benign_rootkit.py
#!/usr/bin/env python3
"""
Benign Rootkit Simulator (sandbox-only)
- Create hidden files/processes in a controlled environment
- Demonstrate hiding and cleanup
- For educational lab use ONLY
"""

import os
import json
import time
import tempfile
import threading
from pathlib import Path

# Config
SANDBOX_ENV = "ROOTKIT_SANDBOX"
LOG_FILE = "rk_log.json"

def safety_check():
    if os.environ.get(SANDBOX_ENV) != "1":
        print("[!] ROOTKIT_SANDBOX env not set. Abort for safety.")
        exit(1)

def create_hidden_file(sandbox_dir, name, content=""):
    hidden_path = Path(sandbox_dir) / f".rk_hidden_{name}"
    with open(hidden_path, "w") as f:
        f.write(content)
    print(f"[+] Hidden file created: {hidden_path}")
    return str(hidden_path)

def fake_process(name, duration=10):
    """Simulate a hidden process using thread"""
    print(f"[+] Starting fake process: {name}")
    time.sleep(duration)
    print(f"[+] Fake process {name} finished")

def log_action(sandbox_dir, action, artifact):
    log_path = Path(sandbox_dir) / LOG_FILE
    entry = {"action": action, "artifact": artifact, "ts": time.time()}
    logs = []
    if log_path.exists():
        with open(log_path, "r") as f:
            logs = json.load(f)
    logs.append(entry)
    with open(log_path, "w") as f:
        json.dump(logs, f, indent=2)

def run_simulation(sandbox_dir):
    # Hidden files
    hidden1 = create_hidden_file(sandbox_dir, "file1.txt", "Secret data")
    hidden2 = create_hidden_file(sandbox_dir, "file2.log", "Simulated log")
    log_action(sandbox_dir, "create_file", hidden1)
    log_action(sandbox_dir, "create_file", hidden2)

    # Fake processes
    t1 = threading.Thread(target=fake_process, args=("proc1",5))
    t2 = threading.Thread(target=fake_process, args=("proc2",7))
    t1.start()
    t2.start()
    log_action(sandbox_dir, "start_process", "proc1")
    log_action(sandbox_dir, "start_process", "proc2")
    t1.join()
    t2.join()

def cleanup(sandbox_dir):
    print("[*] Cleaning up sandbox artifacts...")
    for p in Path(sandbox_dir).glob(".rk_hidden_*"):
        p.unlink()
        print(f"[-] Removed hidden file: {p}")
    log_path = Path(sandbox_dir) / LOG_FILE
    if log_path.exists():
        log_path.unlink()
        print(f"[-] Removed log file: {log_path}")

def main():
    safety_check()
    sandbox_dir = input("Enter sandbox dir (must exist): ").strip()
    if not Path(sandbox_dir).exists():
        print("[!] Sandbox directory does not exist.")
        exit(1)
    choice = input("Run simulation or cleanup? (run/cleanup): ").strip().lower()
    if choice == "run":
        run_simulation(sandbox_dir)
    elif choice == "cleanup":
        cleanup(sandbox_dir)
    else:
        print("Invalid option.")

if __name__ == "__main__":
    main()

File: Projects/Offensive/Rootkit practice/benign_rootkit.c
/*
Benign Rootkit Simulator (C version)
- Sandbox-only educational simulation
- Create/remove hidden files
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <unistd.h>

#define SANDBOX_ENV "ROOTKIT_SANDBOX"

int safety_check() {
    char *env = getenv(SANDBOX_ENV);
    if (!env || strcmp(env,"1") != 0) {
        printf("[!] ROOTKIT_SANDBOX not set. Abort.\n");
        return 0;
    }
    return 1;
}

void create_hidden_file(const char *sandbox, const char *name, const char *content) {
    char path[512];
    snprintf(path, sizeof(path), "%s/.rk_hidden_%s", sandbox, name);
    FILE *f = fopen(path, "w");
    if (!f) return;
    fputs(content, f);
    fclose(f);
    printf("[+] Created hidden file: %s\n", path);
}

void cleanup_hidden_files(const char *sandbox) {
    char path[512];
    snprintf(path, sizeof(path), "%s/", sandbox);
    DIR *d = opendir(path);
    if (!d) return;
    struct dirent *entry;
    while ((entry = readdir(d)) != NULL) {
        if (strstr(entry->d_name, ".rk_hidden_")) {
            snprintf(path, sizeof(path), "%s/%s", sandbox, entry->d_name);
            remove(path);
            printf("[-] Removed %s\n", path);
        }
    }
    closedir(d);
}

int main() {
    if (!safety_check()) return 1;
    char sandbox[256];
    printf("Enter sandbox dir: ");
    scanf("%255s", sandbox);

    int choice;
    printf("Run simulation or cleanup? (1=run,2=cleanup): ");
    scanf("%d",&choice);

    if (choice == 1) {
        create_hidden_file(sandbox, "file1.txt", "Secret data");
        create_hidden_file(sandbox, "file2.log", "Simulated log");
    } else if (choice == 2) {
        cleanup_hidden_files(sandbox);
    } else {
        printf("Invalid choice.\n");
    }
    return 0;
}
